SLEEP-1439 Add permission for offline setup zip download

bramj opened the pull request:
Also: Enable task api in case of access to offline setup file.

Without this, the user will see an error when the frontend tries to render the progress bar.

Related JIRA tickets : [SLEEP-1439](https://bluesquare.atlassian.net/browse/SLEEP-1439) (already in Trypelim)

## Print screen / video

![2024-08-21_10-27](https://github.com/user-attachments/assets/997ee462-08c5-4770-a0ea-7744843089c7)

## How to test

- There should be a new permission "Télécharger l'installation hors ligne de l'application mobile" visible in the user management screen
- When you have this new permission, you see this icon appear as a new user action, allowing you to download a mobile app setup-zip as this user:
![2024-08-21_14-35](https://github.com/user-attachments/assets/edea4a35-22a2-491b-a185-0b70cd6880e0)
- If you don't have the permission, you don't see the action on the users

[SLEEP-1439]: https://bluesquare.atlassian.net/browse/SLEEP-1439?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

bramj committed:
SLEEP-1439 Add permission for offline setup zip download

Also: Enable task api in case of access to offline setup file.

Without this, the user will see an error when the frontend tries to
render the progress bar.

beygorghor commented:
@hakifran we need a review here i think please !

hakifran commented:
> @hakifran we need a review here i think please !

I'm reviewing it

hakifran commented on hat/assets/js/apps/Iaso/domains/users/config.js:
```diff
@@ -92,7 +92,10 @@ export const usersTableColumns = ({
                     )}
-                {currentUser.is_superuser && (
+                {userHasPermission(
+                    Permission.MOBILE_APP_OFFLINE_SETUP,
```
This can be achieved by using 
`<DisplayIfUserHasPerm
                    permissions={[Permission.MOBILE_APP_OFFLINE_SETUP]}
                >
                    <ExportMobileAppSetupDialog
                        selectedUser={settings.row.original}
                        titleMessage={MESSAGES.exportMobileAppTitle}
                        params={params}
                        onCreateExport={exportMobileSetup}
                    />
                </DisplayIfUserHasPerm>`
import { DisplayIfUserHasPerm } from '../../components/DisplayIfUserHasPerm.tsx';

bramj replied:
:+1: adapted

hakifran commented on iaso/api/tasks/__init__.py:
```diff
@@ -72,7 +72,7 @@ class TaskSourceViewSet(ModelViewSet):
     """
 
-    permission_classes = [permissions.IsAuthenticated, HasPermission(permission.DATA_TASKS)]  # type: ignore
+    permission_classes = [permissions.IsAuthenticated, HasPermission(permission.DATA_TASKS, permission.MOBILE_APP_OFFLINE_SETUP)]  # type: ignore
```
@bramj I don't understand why this permission is checked here in TaskSourceVIewSet as this is about:
    """Task API

    GET /api/tasks/
    GET /api/tasks/<id>
    PATCH /api/tasks/<id>
    """

bramj replied:
This was needed to enable me to display a progress bar on the zip creation (which polled the `GET /api/tasks/:id` that was just created).

However, you're right that it is not clear, and too permissive, so I've changed the behavior of only the `GET /api/tasks/<id>` endpoint.

Basically: keep the tasks endpoint permissions as is, but allow a user to query the state of a task launched by themselves, even if they don't have the `DATA_TASKS` permission.

IMO it makes sense to allow this, but only on the `retrieve` endpoint and only for tasks created by the user themself.

@hakifran what do you think?

hakifran replied:
@bramj I do think that the user who launch the task should have the  DATA_TASKS, if not don't let him launch the mobile app offline setup. Because on my point of view this can lead to a confusion about permissions. @beygorghor or @madewulf can give their point of view on this. thank you

bramj committed:
SLEEP-1439 Use DisplayIfUserHasPerm helper component

bramj committed:
SLEEP-1439 Improve Task permission to enable loading states...

such as for the generation of the mobile app setup zip.

Basically: keep the tasks endpoint permissions as is, but allow a user
to query the state of a task launched by themselves, even if they don't
have the "DATA_TASKS" permission.

E.g. For the download of the mobile app setup zip, we display a progress
bar. To display this, we fetch the task that was just created. But if
the user doesn't have the DATA_TASKS permission, this doesn't work.
It makes sense to allow this, but only on the `retrieve` endpoint and
only for tasks created by the user themself.
